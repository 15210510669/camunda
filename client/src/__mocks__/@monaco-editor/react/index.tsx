/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import OriginalEditor, {
  useMonaco as originUseMonaco,
} from '@monaco-editor/react';
import {editor} from 'monaco-editor';

class MockModelContentChangedEvent implements editor.IModelContentChangedEvent {
  changes = [];
  eol = '';
  versionId = 0;
  isUndoing = false;
  isRedoing = false;
  isFlush = false;
}

const Editor: typeof OriginalEditor = ({value, onChange}) => {
  return (
    <textarea
      value={value}
      onChange={(event) => {
        onChange?.(event.target.value, new MockModelContentChangedEvent());
      }}
    />
  );
};

const useMonaco: typeof originUseMonaco = () => {
  return null;
};

export {useMonaco};
export default Editor;
