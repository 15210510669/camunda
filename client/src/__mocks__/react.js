/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

export * from 'react';
export {default} from 'react';

const outstandingEffects = [];
const outstandingCleanups = [];

export const useEffect = (fn) => outstandingEffects.push(fn);
export const runLastEffect = () => {
  if (outstandingEffects.length) {
    const cleanup = outstandingEffects.pop()();

    if (cleanup) {
      outstandingCleanups.push(cleanup);
    }
  }
};

export const runLastCleanup = () => {
  if (outstandingCleanups.length) {
    outstandingCleanups.pop()();
  }
};

export const runAllEffects = () => {
  // Effects can cause scheduling of other effects. To prevent infinite loops,
  // we only run effects that are in the queue when runAllEffects was called.
  const numberOfEffects = outstandingEffects.length;

  for (let i = 0; i < numberOfEffects; i++) {
    const cleanup = outstandingEffects.shift()();

    if (cleanup) {
      outstandingCleanups.push(cleanup);
    }
  }
};

beforeEach(() => {
  outstandingEffects.length = 0;
});

export const runAllCleanups = () => {
  const numberOfEffects = outstandingCleanups.length;

  for (let i = 0; i < numberOfEffects; i++) {
    outstandingCleanups.shift()();
  }
};
