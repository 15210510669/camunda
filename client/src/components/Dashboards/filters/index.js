export {default as FiltersEdit} from './FiltersEdit';
export {default as FiltersView} from './FiltersView';
