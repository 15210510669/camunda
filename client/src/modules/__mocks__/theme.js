export {Provider, Consumer, themed} from '../Theme';
