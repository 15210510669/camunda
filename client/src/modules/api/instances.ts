/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {request} from 'modules/request';
import {RequestFilters} from 'modules/utils/filter';

const URL = '/api/process-instances';

type BatchOperationQuery = {
  active?: boolean;
  canceled?: boolean;
  completed?: boolean;
  excludeIds: string[];
  finished?: boolean;
  ids: string[];
  incidents?: boolean;
  running?: boolean;
};

type ProcessInstancesQuery = {
  query: RequestFilters;
  sorting?: {
    sortBy: string;
    sortOrder: 'desc' | 'asc';
  };
  searchAfter?: ReadonlyArray<string>;
  searchBefore?: ReadonlyArray<string>;
  pageSize?: number;
};

type OperationPayload = {
  operationType: OperationEntityType;
  variableName?: string;
  variableScopeId?: string | undefined;
  variableValue?: string;
  incidentId?: string;
};

type VariablePayload = {
  pageSize: number;
  scopeId: string;
  searchAfter?: ReadonlyArray<string>;
  searchAfterOrEqual?: ReadonlyArray<string>;
  searchBefore?: ReadonlyArray<string>;
  searchBeforeOrEqual?: ReadonlyArray<string>;
};

async function fetchProcessInstance(id: ProcessInstanceEntity['id']) {
  return request({url: `${URL}/${id}`});
}

async function fetchProcessInstanceIncidents(id: ProcessInstanceEntity['id']) {
  return request({url: `${URL}/${id}/incidents`});
}

async function fetchProcessInstances({
  payload,
  signal,
}: {
  payload: ProcessInstancesQuery;
  signal?: AbortSignal;
}) {
  return request({
    url: `${URL}`,
    method: 'POST',
    body: payload,
    signal,
  });
}

async function fetchSequenceFlows(
  processInstanceId: ProcessInstanceEntity['id']
) {
  return request({url: `${URL}/${processInstanceId}/sequence-flows`});
}

async function fetchGroupedProcesses() {
  return request({url: '/api/processes/grouped'});
}

async function fetchProcessCoreStatistics() {
  return request({url: `${URL}/core-statistics`});
}

async function fetchProcessInstancesByIds({
  ids,
  signal,
}: {
  ids: ProcessInstanceEntity['id'][];
  signal?: AbortSignal;
}) {
  return fetchProcessInstances({
    payload: {
      pageSize: ids.length,
      query: {
        running: true,
        finished: true,
        active: true,
        incidents: true,
        completed: true,
        canceled: true,
        ids,
      },
    },
    signal,
  });
}

async function fetchProcessInstancesStatistics(payload: any) {
  return request({
    url: `${URL}/statistics`,
    method: 'POST',
    body: payload,
  });
}

/**
 * @param {*} payload object with query params.
 */
async function applyBatchOperation(
  operationType: OperationEntityType,
  query: BatchOperationQuery
) {
  return request({
    url: `${URL}/batch-operation`,
    method: 'POST',
    body: {
      operationType,
      query,
    },
  });
}

/**
 * @param {*} operationType constants specifying the operation to be applied.
 * @param {*} queries object with query params.
 */
async function applyOperation(
  instanceId: ProcessInstanceEntity['id'],
  payload: OperationPayload
) {
  return request({
    url: `${URL}/${instanceId}/operation`,
    method: 'POST',
    body: payload,
  });
}

async function getOperation(batchOperationId: string) {
  return request({url: `/api/operations?batchOperationId=${batchOperationId}`});
}

async function fetchVariables({
  instanceId,
  payload,
  signal,
}: {
  instanceId: ProcessInstanceEntity['id'];
  payload: VariablePayload;
  signal?: AbortSignal;
}) {
  return request({
    url: `${URL}/${instanceId}/variables`,
    method: 'POST',
    body: payload,
    signal,
  });
}

async function fetchVariable(id: VariableEntity['id']) {
  return request({url: `/api/variables/${id}`});
}

export type {VariablePayload};
export {
  fetchProcessInstances,
  fetchProcessInstance,
  fetchProcessInstanceIncidents,
  fetchSequenceFlows,
  fetchGroupedProcesses,
  fetchProcessCoreStatistics,
  fetchProcessInstancesByIds,
  fetchProcessInstancesStatistics,
  applyBatchOperation,
  applyOperation,
  getOperation,
  fetchVariables,
  fetchVariable,
};
