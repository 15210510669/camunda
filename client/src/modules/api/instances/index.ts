/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import {post, get} from 'modules/request';
import {parseFilterForRequest} from 'modules/utils/filter';
import {FILTER_SELECTION} from 'modules/constants';

const URL = '/api/workflow-instances';

export async function fetchWorkflowInstance(id: any) {
  // @ts-expect-error ts-migrate(2554) FIXME: Expected 2-3 arguments, but got 1.
  const response = await get(`${URL}/${id}`);
  return await response.json();
}

export async function fetchWorkflowInstanceIncidents(id: any) {
  // @ts-expect-error ts-migrate(2554) FIXME: Expected 2-3 arguments, but got 1.
  const response = await get(`${URL}/${id}/incidents`);
  return await response.json();
}

export async function fetchWorkflowInstances(options: any) {
  const {firstResult, maxResults, ...payload} = options;
  const url = `${URL}?firstResult=${firstResult}&maxResults=${maxResults}`;

  const response = await post(url, payload);
  return await response.json();
}

export async function fetchSequenceFlows(workflowInstanceId: any) {
  // @ts-expect-error ts-migrate(2554) FIXME: Expected 2-3 arguments, but got 1.
  const response = await get(`${URL}/${workflowInstanceId}/sequence-flows`);
  return await response.json();
}

export async function fetchGroupedWorkflows() {
  try {
    // @ts-expect-error ts-migrate(2554) FIXME: Expected 2-3 arguments, but got 1.
    const response = await get('/api/workflows/grouped');
    return await response.json();
  } catch (e) {
    return [];
  }
}

export async function fetchWorkflowCoreStatistics() {
  try {
    // @ts-expect-error ts-migrate(2554) FIXME: Expected 2-3 arguments, but got 1.
    const response = await get(`${URL}/core-statistics`);
    return {coreStatistics: await response.json()};
  } catch (e) {
    return {error: e, coreStatistics: {}};
  }
}

export async function fetchWorkflowInstancesByIds(ids: any) {
  const payload = parseFilterForRequest({
    ...FILTER_SELECTION.running,
    ...FILTER_SELECTION.finished,
    ids: ids.join(','),
  });

  const options = {
    firstResult: 0,
    maxResults: ids.length,
    ...payload,
  };
  return await fetchWorkflowInstances(options);
}

export async function fetchWorkflowInstancesBySelection(payload: any) {
  let query = payload.queries[0];

  if (query.ids) {
    query = {
      ...payload.queries[0],
      running: true,
      active: true,
      canceled: true,
      completed: true,
      finished: true,
      incidents: true,
    };
  }

  const url = `${URL}?firstResult=${0}&maxResults=${10}`;
  const response = await post(url, {queries: [...payload.queries]});
  return await response.json();
}

export async function fetchWorkflowInstancesStatistics(payload: any) {
  const url = `${URL}/statistics`;
  const response = await post(url, payload);
  return {statistics: await response.json()};
}

/**
 * @param {*} payload object with query params.
 */
export async function applyBatchOperation(operationType: any, query: any) {
  const url = `${URL}/batch-operation`;
  const payload = {operationType, query};

  const response = await post(url, payload);
  return await response.json();
}

/**
 * @param {*} operationType constants specifying the operation to be applied.
 * @param {*} queries object with query params.
 */
export async function applyOperation(instanceId: any, payload: any) {
  const url = `${URL}/${instanceId}/operation`;

  const response = await post(url, payload);
  return await response.json();
}

export async function fetchVariables({instanceId, scopeId}: any) {
  // TODO: API CHANGED - tests will fail
  // @ts-expect-error ts-migrate(2554) FIXME: Expected 2-3 arguments, but got 1.
  const response = await get(
    `${URL}/${instanceId}/variables?scopeId=${scopeId}`
  );
  return await response.json();
}
