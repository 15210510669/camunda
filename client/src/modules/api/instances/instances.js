import {post, get} from 'modules/request';

const URL = '/api/workflow-instances';

export async function fetchWorkflowInstance(id) {
  const response = await get(`${URL}/${id}`);
  return await response.json();
}

export async function fetchWorkflowInstances(options) {
  const {firstResult, maxResults, ...payload} = options;
  const url = `${URL}?firstResult=${firstResult}&maxResults=${maxResults}`;
  const response = await post(url, payload);

  return await response.json();
}

export async function fetchGroupedWorkflowInstances() {
  try {
    const response = await get('/api/workflows/grouped');
    return await response.json();
  } catch (e) {
    return [];
  }
}

export async function fetchWorkflowInstancesCount(payload) {
  const url = `${URL}?firstResult=${0}&maxResults=${0}`;
  const response = await post(url, {
    queries: [{...payload}]
  });
  const resJson = await response.json();
  return resJson.totalCount;
}

export async function fetchWorkflowInstanceBySelection(payload) {
  let query = payload.queries[0];

  if (query.ids) {
    query = {
      ...payload.queries[0],
      running: true,
      active: true,
      canceled: true,
      completed: true,
      finished: true,
      incidents: true
    };
  }

  const url = `${URL}?firstResult=${0}&maxResults=${10}`;
  const response = await post(url, {queries: [...payload.queries]});
  return await response.json();
}

export async function fetchWorkflowInstancesStatistics(payload) {
  const url = `${URL}/statistics`;
  const response = await post(url, {queries: [...payload.queries]});
  return await response.json();
}

/**
 * @param {*} operationType constants specifying the operation to be applied.
 * @param {*} queries object with query params.
 */
export async function applyOperation(operationType, queries) {
  const url = `${URL}/operation`;
  const payload = {operationType, queries};

  await post(url, payload);
}
