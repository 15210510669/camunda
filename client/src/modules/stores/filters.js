/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import {
  observable,
  decorate,
  action,
  computed,
  observe,
  reaction,
  when,
} from 'mobx';
import {DEFAULT_FIRST_ELEMENT, DEFAULT_SORTING} from 'modules/constants';
import {formatGroupedWorkflows} from 'modules/utils/instance';
import {
  parseFilterForRequest,
  parseQueryString,
  getWorkflowByVersion,
  getFilterWithWorkflowIds,
  decodeFields,
} from 'modules/utils/filter';
import {fetchGroupedWorkflows} from 'modules/api/instances';
import {isEqual, isEmpty} from 'lodash';
import {sanitizeFilter} from './utils/sanitizeFilter';
import {setBrowserUrl} from './utils/setBrowserUrl';

const DEFAULT_STATE = {
  filter: {},
  page: 1,
  entriesPerPage: 0,
  prevEntriesPerPage: 0,
  firstElement: DEFAULT_FIRST_ELEMENT,
  sorting: DEFAULT_SORTING,
  groupedWorkflows: {},
  batchOperationId: '',
  history: null,
  location: null,
  isInitialLoadComplete: false,
};

class Filters {
  state = {...DEFAULT_STATE};
  filterObserveDisposer = null;
  locationReactionDisposer = null;

  async init() {
    this.setGroupedWorkflows(
      formatGroupedWorkflows(await fetchGroupedWorkflows())
    );

    when(
      () => this.state.location !== null,
      () => {
        this.setFilterFromUrl();
        const queryParams = parseQueryString(this.state.location.search);

        if (!isEqual(this.state.filter, queryParams.filter)) {
          setBrowserUrl(
            this.state.history,
            this.state.location,
            this.state.filter,
            this.state.groupedWorkflows,
            queryParams.name
          );
        }

        this.completeInitialLoad();
      }
    );

    this.locationReactionDisposer = reaction(
      () => this.state.location,
      () => {
        this.setFilterFromUrl();
      }
    );

    this.filterObserveDisposer = observe(this.state, 'filter', (change) => {
      if (isEqual(this.state.filter, change.oldValue)) {
        return;
      }

      setBrowserUrl(
        this.state.history,
        this.state.location,
        this.state.filter,
        this.state.groupedWorkflows
      );
      this.setPage(1);
    });
  }

  completeInitialLoad = () => {
    this.state.isInitialLoadComplete = true;
  };

  setUrlParameters = (history, location) => {
    this.state.history = history;
    this.state.location = location;
  };

  setGroupedWorkflows = (groupedWorkflows) => {
    this.state.groupedWorkflows = groupedWorkflows;
  };

  setFilter = (filter) => {
    if (
      !filter.canceled &&
      !filter.completed &&
      this.state.sorting.sortBy === 'endDate'
    ) {
      this.setSorting(DEFAULT_SORTING);
    }

    this.state.filter = filter;
  };

  setFilterFromUrl() {
    const filterFromURL = sanitizeFilter(
      parseQueryString(this.state.location.search).filter,
      this.state.groupedWorkflows
    );

    if (!isEqual(this.state.filter, filterFromURL)) {
      this.state.filter = filterFromURL;
    }
  }

  setPage(page) {
    this.state.page = page;
  }

  setSorting(sorting) {
    this.state.sorting = sorting;
  }

  setEntriesPerPage(entriesPerPage) {
    this.state.prevEntriesPerPage = this.state.entriesPerPage;
    this.state.entriesPerPage = entriesPerPage;
  }

  getFiltersPayload() {
    const {filter, groupedWorkflows} = this.state;
    return parseFilterForRequest(
      getFilterWithWorkflowIds(filter, groupedWorkflows)
    );
  }

  get decodedFilters() {
    return decodeFields(this.state.filter);
  }

  get firstElement() {
    return (this.state.page - 1) * this.state.entriesPerPage;
  }

  get isNoVersionSelected() {
    return this.state.filter?.version === 'all';
  }

  get isNoWorkflowSelected() {
    return !this.state.filter?.workflow;
  }

  get workflow() {
    const {groupedWorkflows, filter} = this.state;
    return getWorkflowByVersion(
      groupedWorkflows[filter.workflow],
      filter.version
    );
  }

  get workflowName() {
    const workflow = !isEmpty(this.workflow)
      ? this.workflow
      : this.state.groupedWorkflows[this.state.filter.workflow];

    return workflow ? workflow.name || workflow.bpmnProcessId : 'Workflow';
  }

  reset = () => {
    this.state = {...DEFAULT_STATE};

    if (this.filterObserveDisposer !== null) {
      this.filterObserveDisposer();
    }
    if (this.locationReactionDisposer !== null) {
      this.locationReactionDisposer();
    }
  };
}

decorate(Filters, {
  state: observable,
  setFilterFromUrl: action,
  setFilter: action,
  setPage: action,
  setSorting: action,
  setEntriesPerPage: action,
  reset: action,
  firstElement: computed,
  decodedFilters: computed,
  isNoVersionSelected: computed,
  isNoWorkflowSelected: computed,
  workflow: computed,
  workflowName: computed,
  setGroupedWorkflows: action,
  completeInitialLoad: action,
  setUrlParameters: action,
});

export const filters = new Filters();
