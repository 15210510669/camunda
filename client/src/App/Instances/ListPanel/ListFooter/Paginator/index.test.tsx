/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React from 'react';
import {render, screen, fireEvent} from '@testing-library/react';
import {ThemeProvider} from 'modules/theme/ThemeProvider';
import {filtersStore} from 'modules/stores/filters';
import {instancesStore} from 'modules/stores/instances';
import {Paginator} from './index';

const mockInstances = [
  {
    id: '2251799813685625',
    workflowId: '2251799813685623',
    workflowName: 'Without Incidents Process',
    workflowVersion: 1,
    startDate: '2020-11-19T08:14:05.406+0000',
    endDate: null,
    state: 'ACTIVE',
    bpmnProcessId: 'withoutIncidentsProcess',
    hasActiveOperation: false,
    operations: [],
  } as const,
  {
    id: '2251799813685627',
    workflowId: '2251799813685623',
    workflowName: 'Without Incidents Process',
    workflowVersion: 1,
    startDate: '2020-11-19T08:14:05.490+0000',
    endDate: null,
    state: 'ACTIVE',
    bpmnProcessId: 'withoutIncidentsProcess',
    hasActiveOperation: false,
    operations: [],
  } as const,
];

describe('Paginator', () => {
  afterEach(() => {
    filtersStore.reset();
    instancesStore.reset();
  });

  it('should show the first five pages when on first page', () => {
    filtersStore.setEntriesPerPage(10);
    filtersStore.setPage(1);
    instancesStore.setInstances({
      filteredInstancesCount: 150,
      workflowInstances: mockInstances,
    });

    render(<Paginator maxPage={15} />, {wrapper: ThemeProvider});

    [1, 2, 3, 4, 5, 15].forEach((pageNumber) => {
      expect(
        screen.getByRole('button', {name: `Page ${pageNumber}`})
      ).toBeInTheDocument();
    });

    expect(screen.getByText('…')).toBeInTheDocument();
    expect(
      screen.queryByRole('button', {name: 'Page 6'})
    ).not.toBeInTheDocument();
  });

  it('should show the last five pages when on last page', () => {
    filtersStore.setEntriesPerPage(10);
    filtersStore.setPage(15);
    instancesStore.setInstances({
      filteredInstancesCount: 150,
      workflowInstances: mockInstances,
    });

    render(<Paginator maxPage={15} />, {wrapper: ThemeProvider});
    [15, 14, 13, 12, 11, 1].forEach((pageNumber) => {
      expect(
        screen.getByRole('button', {name: `Page ${pageNumber}`})
      ).toBeInTheDocument();
    });
    expect(screen.getByText('…')).toBeInTheDocument();
    expect(
      screen.queryByRole('button', {name: 'Page 10'})
    ).not.toBeInTheDocument();
  });

  it('should show ellipsis on both sides in the middle', () => {
    filtersStore.setEntriesPerPage(5);
    filtersStore.setPage(10);
    instancesStore.setInstances({
      filteredInstancesCount: 150,
      workflowInstances: mockInstances,
    });

    render(<Paginator maxPage={20} />, {wrapper: ThemeProvider});
    expect(screen.getAllByText('…').length).toBe(2);
  });

  it('should not show ellipsis when first page is just barely out of range', () => {
    filtersStore.setEntriesPerPage(10);
    filtersStore.setPage(1);
    instancesStore.setInstances({
      filteredInstancesCount: 20,
      workflowInstances: mockInstances,
    });
    render(<Paginator maxPage={2} />, {wrapper: ThemeProvider});
    expect(screen.queryByText('…')).not.toBeInTheDocument();
  });

  it('should navigate between pages', () => {
    filtersStore.setEntriesPerPage(10);
    filtersStore.setPage(1);
    instancesStore.setInstances({
      filteredInstancesCount: 30,
      workflowInstances: mockInstances,
    });
    render(<Paginator maxPage={3} />, {wrapper: ThemeProvider});

    expect(screen.getByRole('button', {name: 'First page'})).toBeDisabled();
    expect(screen.getByRole('button', {name: 'Previous page'})).toBeDisabled();
    expect(screen.getByRole('button', {name: 'Last page'})).toBeEnabled();
    expect(screen.getByRole('button', {name: 'Next page'})).toBeEnabled();
    expect(filtersStore.state.page).toBe(1);

    fireEvent.click(screen.getByRole('button', {name: 'Next page'}));
    ['First page', 'Previous page', 'Last page', 'Next page'].forEach((name) =>
      expect(screen.getByRole('button', {name})).toBeEnabled()
    );
    expect(filtersStore.state.page).toBe(2);

    fireEvent.click(screen.getByRole('button', {name: 'Next page'}));
    expect(screen.getByRole('button', {name: 'First page'})).toBeEnabled();
    expect(screen.getByRole('button', {name: 'Previous page'})).toBeEnabled();
    expect(screen.getByRole('button', {name: 'Last page'})).toBeDisabled();
    expect(screen.getByRole('button', {name: 'Next page'})).toBeDisabled();
    expect(filtersStore.state.page).toBe(3);

    fireEvent.click(screen.getByRole('button', {name: 'First page'}));
    expect(screen.getByRole('button', {name: 'First page'})).toBeDisabled();
    expect(screen.getByRole('button', {name: 'Previous page'})).toBeDisabled();
    expect(screen.getByRole('button', {name: 'Last page'})).toBeEnabled();
    expect(screen.getByRole('button', {name: 'Next page'})).toBeEnabled();
    expect(filtersStore.state.page).toBe(1);

    fireEvent.click(screen.getByRole('button', {name: 'Last page'}));
    expect(screen.getByRole('button', {name: 'First page'})).toBeEnabled();
    expect(screen.getByRole('button', {name: 'Previous page'})).toBeEnabled();
    expect(screen.getByRole('button', {name: 'Last page'})).toBeDisabled();
    expect(screen.getByRole('button', {name: 'Next page'})).toBeDisabled();
    expect(filtersStore.state.page).toBe(3);

    fireEvent.click(screen.getByRole('button', {name: 'Previous page'}));
    ['First page', 'Previous page', 'Last page', 'Next page'].forEach((name) =>
      expect(screen.getByRole('button', {name})).toBeEnabled()
    );
    expect(filtersStore.state.page).toBe(2);
  });
});
