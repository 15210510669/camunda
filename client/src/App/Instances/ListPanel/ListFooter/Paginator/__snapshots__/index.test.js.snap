// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not show ellipsis when first page is just barely out of range 1`] = `
<styled.div>
  <Themed(styled.button)
    aria-label="First page"
    disabled={false}
    onClick={[Function]}
    title="First page"
    withIcon={true}
  >
    <ForwardRef(SvgLeftBar) />
  </Themed(styled.button)>
  <Themed(styled.button)
    aria-label="Previous page"
    disabled={false}
    onClick={[Function]}
    title="Previous page"
    withIcon={true}
  >
    <ForwardRef(SvgLeft) />
  </Themed(styled.button)>
  <Themed(styled.button)
    aria-label="First page"
    onClick={[Function]}
    title="First page"
  >
    1
  </Themed(styled.button)>
  <Themed(styled.button)
    active={false}
    aria-label="Page 2"
    key="2"
    onClick={[Function]}
    title="Page 2"
  >
    2
  </Themed(styled.button)>
  <Themed(styled.button)
    active={false}
    aria-label="Page 3"
    key="3"
    onClick={[Function]}
    title="Page 3"
  >
    3
  </Themed(styled.button)>
  <Themed(styled.button)
    active={true}
    aria-label="Page 4"
    key="4"
    onClick={[Function]}
    title="Page 4"
  >
    4
  </Themed(styled.button)>
  <Themed(styled.button)
    active={false}
    aria-label="Page 5"
    key="5"
    onClick={[Function]}
    title="Page 5"
  >
    5
  </Themed(styled.button)>
  <Themed(styled.button)
    active={false}
    aria-label="Page 6"
    key="6"
    onClick={[Function]}
    title="Page 6"
  >
    6
  </Themed(styled.button)>
  <styled.div>
    …
  </styled.div>
  <Themed(styled.button)
    aria-label="Page 10"
    onClick={[Function]}
    title="Page 10"
  >
    10
  </Themed(styled.button)>
  <Themed(styled.button)
    aria-label="Next page"
    data-test="next-page"
    disabled={false}
    onClick={[Function]}
    title="Next page"
    withIcon={true}
  >
    <ForwardRef(SvgRight) />
  </Themed(styled.button)>
  <Themed(styled.button)
    aria-label="Last page"
    disabled={false}
    onClick={[Function]}
    title="Last page"
    withIcon={true}
  >
    <ForwardRef(SvgRightBar) />
  </Themed(styled.button)>
</styled.div>
`;

exports[`should show ellipsis on both sides in the middle 1`] = `
<styled.div>
  <Themed(styled.button)
    aria-label="First page"
    disabled={false}
    onClick={[Function]}
    title="First page"
    withIcon={true}
  >
    <ForwardRef(SvgLeftBar) />
  </Themed(styled.button)>
  <Themed(styled.button)
    aria-label="Previous page"
    disabled={false}
    onClick={[Function]}
    title="Previous page"
    withIcon={true}
  >
    <ForwardRef(SvgLeft) />
  </Themed(styled.button)>
  <Themed(styled.button)
    aria-label="First page"
    onClick={[Function]}
    title="First page"
  >
    1
  </Themed(styled.button)>
  <styled.div>
    …
  </styled.div>
  <Themed(styled.button)
    active={false}
    aria-label="Page 9"
    key="9"
    onClick={[Function]}
    title="Page 9"
  >
    9
  </Themed(styled.button)>
  <Themed(styled.button)
    active={false}
    aria-label="Page 10"
    key="10"
    onClick={[Function]}
    title="Page 10"
  >
    10
  </Themed(styled.button)>
  <Themed(styled.button)
    active={true}
    aria-label="Page 11"
    key="11"
    onClick={[Function]}
    title="Page 11"
  >
    11
  </Themed(styled.button)>
  <Themed(styled.button)
    active={false}
    aria-label="Page 12"
    key="12"
    onClick={[Function]}
    title="Page 12"
  >
    12
  </Themed(styled.button)>
  <Themed(styled.button)
    active={false}
    aria-label="Page 13"
    key="13"
    onClick={[Function]}
    title="Page 13"
  >
    13
  </Themed(styled.button)>
  <styled.div>
    …
  </styled.div>
  <Themed(styled.button)
    aria-label="Page 20"
    onClick={[Function]}
    title="Page 20"
  >
    20
  </Themed(styled.button)>
  <Themed(styled.button)
    aria-label="Next page"
    data-test="next-page"
    disabled={false}
    onClick={[Function]}
    title="Next page"
    withIcon={true}
  >
    <ForwardRef(SvgRight) />
  </Themed(styled.button)>
  <Themed(styled.button)
    aria-label="Last page"
    disabled={false}
    onClick={[Function]}
    title="Last page"
    withIcon={true}
  >
    <ForwardRef(SvgRightBar) />
  </Themed(styled.button)>
</styled.div>
`;

exports[`should show the first five pages when on first page 1`] = `
<styled.div>
  <Themed(styled.button)
    aria-label="First page"
    disabled={true}
    onClick={[Function]}
    title="First page"
    withIcon={true}
  >
    <ForwardRef(SvgLeftBar) />
  </Themed(styled.button)>
  <Themed(styled.button)
    aria-label="Previous page"
    disabled={true}
    onClick={[Function]}
    title="Previous page"
    withIcon={true}
  >
    <ForwardRef(SvgLeft) />
  </Themed(styled.button)>
  <Themed(styled.button)
    active={true}
    aria-label="Page 1"
    key="1"
    onClick={[Function]}
    title="Page 1"
  >
    1
  </Themed(styled.button)>
  <Themed(styled.button)
    active={false}
    aria-label="Page 2"
    key="2"
    onClick={[Function]}
    title="Page 2"
  >
    2
  </Themed(styled.button)>
  <Themed(styled.button)
    active={false}
    aria-label="Page 3"
    key="3"
    onClick={[Function]}
    title="Page 3"
  >
    3
  </Themed(styled.button)>
  <Themed(styled.button)
    active={false}
    aria-label="Page 4"
    key="4"
    onClick={[Function]}
    title="Page 4"
  >
    4
  </Themed(styled.button)>
  <Themed(styled.button)
    active={false}
    aria-label="Page 5"
    key="5"
    onClick={[Function]}
    title="Page 5"
  >
    5
  </Themed(styled.button)>
  <styled.div>
    …
  </styled.div>
  <Themed(styled.button)
    aria-label="Page 15"
    onClick={[Function]}
    title="Page 15"
  >
    15
  </Themed(styled.button)>
  <Themed(styled.button)
    aria-label="Next page"
    data-test="next-page"
    disabled={false}
    onClick={[Function]}
    title="Next page"
    withIcon={true}
  >
    <ForwardRef(SvgRight) />
  </Themed(styled.button)>
  <Themed(styled.button)
    aria-label="Last page"
    disabled={false}
    onClick={[Function]}
    title="Last page"
    withIcon={true}
  >
    <ForwardRef(SvgRightBar) />
  </Themed(styled.button)>
</styled.div>
`;

exports[`should show the last five pages when on last page 1`] = `
<styled.div>
  <Themed(styled.button)
    aria-label="First page"
    disabled={false}
    onClick={[Function]}
    title="First page"
    withIcon={true}
  >
    <ForwardRef(SvgLeftBar) />
  </Themed(styled.button)>
  <Themed(styled.button)
    aria-label="Previous page"
    disabled={false}
    onClick={[Function]}
    title="Previous page"
    withIcon={true}
  >
    <ForwardRef(SvgLeft) />
  </Themed(styled.button)>
  <Themed(styled.button)
    aria-label="First page"
    onClick={[Function]}
    title="First page"
  >
    1
  </Themed(styled.button)>
  <styled.div>
    …
  </styled.div>
  <Themed(styled.button)
    active={false}
    aria-label="Page 7"
    key="7"
    onClick={[Function]}
    title="Page 7"
  >
    7
  </Themed(styled.button)>
  <Themed(styled.button)
    active={false}
    aria-label="Page 8"
    key="8"
    onClick={[Function]}
    title="Page 8"
  >
    8
  </Themed(styled.button)>
  <Themed(styled.button)
    active={false}
    aria-label="Page 9"
    key="9"
    onClick={[Function]}
    title="Page 9"
  >
    9
  </Themed(styled.button)>
  <Themed(styled.button)
    active={true}
    aria-label="Page 10"
    key="10"
    onClick={[Function]}
    title="Page 10"
  >
    10
  </Themed(styled.button)>
  <Themed(styled.button)
    active={false}
    aria-label="Page 11"
    key="11"
    onClick={[Function]}
    title="Page 11"
  >
    11
  </Themed(styled.button)>
  <Themed(styled.button)
    aria-label="Next page"
    data-test="next-page"
    disabled={false}
    onClick={[Function]}
    title="Next page"
    withIcon={true}
  >
    <ForwardRef(SvgRight) />
  </Themed(styled.button)>
  <Themed(styled.button)
    aria-label="Last page"
    disabled={false}
    onClick={[Function]}
    title="Last page"
    withIcon={true}
  >
    <ForwardRef(SvgRightBar) />
  </Themed(styled.button)>
</styled.div>
`;
